src/lib/pickles/wrap_verifier.ml

```diff
@@ -836,15 +887,50 @@ struct
               with
               | _ :: Some _ :: _, _ | _, Some _ ->
                   let joint_combiner = sample_scalar () in
-                  absorb_sorted_first_part () ;
-                  absorb_sorted_second_part () ;
+                  absorb_sorted_1 sponge ;
+                  absorb_sorted_2_to_4 () ;
+                  absorb_sorted_5 () ;
                   Types.Opt.Some joint_combiner
               | _ :: None :: _, None ->
-                  absorb_sorted_first_part () ;
-                  absorb_sorted_second_part () ;
+                  absorb_sorted_1 sponge ;
+                  absorb_sorted_2_to_4 () ;
+                  absorb_sorted_5 () ;
                   Types.Opt.Some { inner = Field.zero }
-              | _ ->
-                  failwith "TODO" )
+              | _ :: Maybe (b1, _) :: _, Maybe (b2, _) ->
+                  let b = Boolean.(b1 ||| b2) in
+                  let sponge2 = Opt.copy sponge in
+                  let joint_combiner_if_true =
+                    let joint_combiner = sample_scalar () in
+                    absorb_sorted_1 sponge ; joint_combiner
+                  in
+                  let joint_combiner_if_false : Scalar_challenge.t =
+                    absorb_sorted_1 sponge2 ; { inner = Field.zero }
+                  in
+                  Opt.recombine b ~original_sponge:sponge2 sponge ;
+                  absorb_sorted_2_to_4 () ;
+                  absorb_sorted_5 () ;
+                  Types.Opt.Some
+                    { inner =
+                        Field.if_ b ~then_:joint_combiner_if_true.inner
+                          ~else_:joint_combiner_if_false.inner
+                    }
+              | _ :: Maybe (b, _) :: _, _ | _, Maybe (b, _) ->
+                  let sponge2 = Opt.copy sponge in
+                  let joint_combiner_if_true =
+                    let joint_combiner = sample_scalar () in
+                    absorb_sorted_1 sponge ; joint_combiner
+                  in
+                  let joint_combiner_if_false : Scalar_challenge.t =
+                    absorb_sorted_1 sponge2 ; { inner = Field.zero }
+                  in
+                  Opt.recombine b ~original_sponge:sponge2 sponge ;
+                  absorb_sorted_2_to_4 () ;
+                  absorb_sorted_5 () ;
+                  Types.Opt.Some
+                    { inner =
+                        Field.if_ b ~then_:joint_combiner_if_true.inner
+                          ~else_:joint_combiner_if_false.inner
+                    } )
```

Comment:
901-916 and 918-933 look like the same code. Make a function with `Boolean.t` parameter?